File: goblinframework-core/src/main/java/org/goblinframework/core/serialization/java/JavaSerializer.java
package org.goblinframework.core.serialization.java;

import org.apache.commons.lang3.SerializationUtils;
import org.goblinframework.api.common.Singleton;
import org.goblinframework.core.exception.GoblinSerializationException;
import org.goblinframework.core.serialization.Serializer;
import org.goblinframework.core.serialization.SerializerMode;
import org.jetbrains.annotations.NotNull;

import java.io.InputStream;
import java.io.OutputStream;
import java.io.Serializable;

@Singleton
final public class JavaSerializer implements Serializer {

  public static final JavaSerializer INSTANCE = new JavaSerializer();

  private JavaSerializer() {
  }

  @NotNull
  @Override
  public SerializerMode mode() {
    return SerializerMode.JAVA;
  }

  @Override
  public void serialize(@NotNull Object obj, @NotNull OutputStream outStream) {
    if (!(obj instanceof Serializable)) {
      throw GoblinSerializationException.requiredSerializable(obj);
    }
    try {
      SerializationUtils.serialize((Serializable) obj, outStream);
    } catch (Exception ex) {
      throw new GoblinSerializationException(ex);
    }
  }

  @NotNull
  @Override
  public byte[] serialize(@NotNull Object obj) {
    if (!(obj instanceof Serializable)) {
      throw GoblinSerializationException.requiredSerializable(obj);
    }
    try {
      return SerializationUtils.serialize((Serializable) obj);
    } catch (Exception ex) {
      throw new GoblinSerializationException(ex);
    }
  }

  @NotNull
  @Override
  public Object deserialize(@NotNull InputStream inStream) {
    try {
      return SerializationUtils.deserialize(inStream);
    } catch (Exception ex) {
      throw new GoblinSerializationException(ex);
    }
  }

  @NotNull
  @Override
  public Object deserialize(@NotNull byte[] bs) {
    try {
      return SerializationUtils.deserialize(bs);
    } catch (Exception ex) {
      throw new GoblinSerializationException(ex);
    }
  }
}

File: goblinframework-core/src/main/java/org/goblinframework/core/serialization/java/JavaSerializer.kt
package org.goblinframework.core.serialization.java

import org.apache.commons.lang3.SerializationUtils
import org.goblinframework.core.serialization.Serializer
import java.io.InputStream
import java.io.OutputStream
import java.io.Serializable

class JavaSerializer : Serializer {

  override fun id(): Byte {
    return Serializer.JAVA
  }

  override fun serialize(obj: Any, outStream: OutputStream) {
    if (obj !is Serializable) {
      throw UnsupportedOperationException("Serializable is required")
    }
    SerializationUtils.serialize(obj, outStream)
  }

  override fun serialize(obj: Any): ByteArray {
    if (obj !is Serializable) {
      throw UnsupportedOperationException("Serializable is required")
    }
    return SerializationUtils.serialize(obj)
  }

  override fun deserialize(inStream: InputStream): Any {
    return SerializationUtils.deserialize(inStream)
  }

  override fun deserialize(bs: ByteArray): Any {
    return SerializationUtils.deserialize(bs)
  }
}